File: medium/46. Permutations.java
class Solution {
    //TC : O(n*n!)
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        dfs(result, new ArrayList<Integer>(), nums);
        return result;
    }
    
    public void dfs(List<List<Integer>> result, List<Integer> list, int[] nums) {
        if(list.size() == nums.length) {
            result.add(new ArrayList<>(list));
        }
        
        for(int i=0;i<nums.length;i++){
            if(list.contains(nums[i])) continue;
            list.add(nums[i]);
            dfs(result, list, nums);
            list.remove(list.size()-1);
        }
    }
}

File: medium/46. Permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        '''
        nums <-- only consist of the numbers? -> negative,positive,0 / duplicates?
        nums.length -> can be blank?
        
        output = 2d array
        
        TC:O(n!)
        SC : O(n)
        '''
        def dfs(path, max_len)  :
            if len(path) == max_len :
                res.append(path)
            
            for x in nums :
                if x not in path :
                    dfs(path+[x],  max_len)
            
        res = []
        
        for x in nums :
            dfs([x], len(nums))
        return res
    
    
    
        
        
